jupyterhub/app.py

```diff
@@ -1975,25 +1974,42 @@ async def init_groups(self):
                 group.users.append(user)
         db.commit()
 
-    async def init_roles(self):
+    async def init_role_creation(self):
         """Load default and predefined roles into the database"""
-        db = self.db
-        # tokens are added separately
-        role_bearers = ['users', 'services', 'groups']
-
-        # load default roles
         self.log.debug('Loading default roles to database')
         default_roles = roles.get_default_roles()
-        for role in default_roles:
-            roles.create_role(db, role)
+        default_role_names = [r['name'] for r in default_roles]
+        init_roles = default_roles + self.load_roles
+        if not orm.Role.find(self.db, name='admin'):
+            self._rbac_upgrade = True
+        for role in self.db.query(orm.Role):
+            if role.name not in default_role_names:
+                self.db.delete(role)
+        self.db.commit()
+        for role in init_roles:
+            roles.create_role(self.db, role)
+
+    async def init_role_assignment(self):
+        # tokens are added separately
+        role_bearers = ['users', 'services', 'groups']
 
+        db = self.db
         # load predefined roles from config file
+        for role_spec in self.load_roles:
+            if role_spec['name'] == 'admin' and self.admin_users:
+                app_log.info(
+                    "Extending admin role assignment with config admin users: %s",
+                    str(self.admin_users),
+                )
+                role_spec['users'].extend(self.admin_users)
```

Comment:
Handle `'users'` being undefined in the original role spec (it could e.g. have only services):


```suggestion
                role_spec.setdefault('users', [])
                role_spec['users'].extend(self.admin_users)
```

Reply:
Good to know, thanks. I forgot to take the `service` spec into account at all, actually. I'll push a fix for that as well